// sequency/lib/distinct.js
Object.defineProperty(exports, "__esModule", { value: true });
var Sequence_1 = require("./Sequence");
var DistinctIterator = (function () {
    function DistinctIterator(iterator) {
        this.iterator = iterator;
        this.items = [];
        this.nextItem = undefined;
        this.done = false;
    }
    DistinctIterator.prototype.hasNext = function () {
        this.processNext();
        return !this.done;
    };
    DistinctIterator.prototype.next = function () {
        this.processNext();
        var result = this.nextItem;
        this.nextItem = undefined;
        return result;
    };
    DistinctIterator.prototype.processNext = function () {
        if (this.nextItem !== undefined || this.done) {
            return;
        }
        while (this.iterator.hasNext()) {
            var item = this.iterator.next();
            if (this.items.indexOf(item) < 0) {
                this.nextItem = item;
                this.items.push(item);
                return;
            }
        }
        this.done = true;
    };
    return DistinctIterator;
}());
/**
 * Returns a new sequence which discards all duplicate elements.
 *
 * @returns {Sequence<T>}
 */
function distinct() {
    return new Sequence_1.default(new DistinctIterator(this.iterator));
}
exports.default = distinct;
