// sequency/lib/filter.js
Object.defineProperty(exports, "__esModule", { value: true });
var Sequence_1 = require("./Sequence");
var FilterIterator = (function () {
    function FilterIterator(predicate, iterator) {
        this.predicate = predicate;
        this.iterator = iterator;
        this.done = false;
    }
    FilterIterator.prototype.hasNext = function () {
        this.processNext();
        return !this.done;
    };
    FilterIterator.prototype.next = function () {
        this.processNext();
        var result = this.nextItem;
        this.nextItem = undefined;
        return result;
    };
    FilterIterator.prototype.processNext = function () {
        if (this.done || this.nextItem !== undefined) {
            return;
        }
        while (this.iterator.hasNext()) {
            var item = this.iterator.next();
            if (this.predicate(item)) {
                this.nextItem = item;
                return;
            }
        }
        this.done = true;
    };
    return FilterIterator;
}());
/**
 * Returns a new sequence consisting of all elements that match the given `predicate`.
 *
 * @param {(T) => boolean} predicate
 * @returns {Sequence<T>}
 */
function filter(predicate) {
    return new Sequence_1.default(new FilterIterator(predicate, this.iterator));
}
exports.default = filter;
